plugins/automation/app/services/discourse_automation/update.rb

```diff
@@ -0,0 +1,183 @@
+    automation.fields.destroy_all
+    context[:fields_cleared] = true
+    context[:force_disable] = true
+  end
+
+  def handle_script_change(automation:, params:)
+    return if params.script.blank?
+    return if automation.script == params.script
+
+    automation.fields.destroy_all
+    context[:fields_cleared] = true
+    context[:force_disable] = true
+    context[:clear_trigger] = true
+  end
+
+  def update_automation(automation:, params:, guardian:)
+    automation.perform_required_fields_validation = true
+
+    attributes = { last_updated_by_id: guardian.user.id }
+    attributes[:name] = params.name if params.name.present?
+    attributes[:script] = params.script if params.script.present?
+    attributes[:trigger] = context[:forced_trigger] || params.trigger if params.trigger.present? ||
+      context[:forced_trigger].present?
+    attributes[:trigger] = nil if context[:clear_trigger]
+    attributes[:enabled] = params.enabled unless params.enabled.nil?
+    attributes[:enabled] = false if context[:force_disable]
+
+    # Update fields if not cleared and fields provided
+    if !context[:fields_cleared] && params.fields.present?
+      Array(params.fields)
+        .reject(&:empty?)
+        .each do |field|
+          automation.upsert_field!(
+            field[:name],
+            field[:component],
+            field[:metadata],
+            target: field[:target],
+          )
+        end
+    end
+
+    if context[:clear_trigger]
+      automation.assign_attributes(attributes)
+      automation.save!(validate: false)
+    else
+      automation.update!(attributes)
+    end
+
+    automation.reload
+  end
+
+  def log_action(automation:, guardian:)
+    previous_state = context[:previous_state]
+    changes = {}
+
+    %i[name script trigger enabled].each do |attr|
+      current_value = automation.public_send(attr)
+      previous_value = previous_state[attr]
+      if current_value != previous_value
+        changes[attr] = "#{format_value(previous_value)} → #{format_value(current_value)}"
+      end
+    end
+
+    current_fields = automation.serialized_fields
+    previous_fields = previous_state[:fields]
+
+    all_field_names = (current_fields.keys + previous_fields.keys).uniq.sort
+
+    all_field_names.each do |field_name|
+      current_field = current_fields[field_name]
+      previous_field = previous_fields[field_name]
+
+      next if current_field == previous_field
+      next if field_empty?(previous_field) && field_empty?(current_field)
+
+      changes[
+        field_name
+      ] = "#{format_field_value(previous_field)} → #{format_field_value(current_field)}"
+    end
+
+    return if changes.empty?
+
+    details = { id: automation.id }
+    # Only show name as identifier if it wasn't changed
+    details[:name] = automation.name unless changes.key?(:name)
+    details.merge!(changes)
+
+    StaffActionLogger.new(guardian.user).log_custom("update_automation", **details)
+  end
+
+  def format_value(value)
```

Comment:
Avoid utility methods in a service, see https://meta.discourse.org/t/using-service-objects-in-discourse/333641#p-1635143-best-practices-and-guidelines-61.
There’s a lot of logic in the `update_automation` and `log_action`, it should probably be extracted to dedicated actions.